### PR TITLE
Add new C++ driven proof machine fuzzer

### PR DESCRIPTION
**Pull request by PlasmaPower**
This also features a new machine serialization format aimed to be suitable for fuzzing, and fixes various issues found by the fuzzer.

### CODE REVIEW

File: packages/arb-avm-cpp/CMakeLists.txt [hunk 1]
```diff
@@ -57,7 +57,25 @@ set(AVM_VERSION_STRING ${AVM_VERSION_MAJOR}.${AVM_VERSION_MINOR}.${AVM_VERSION_P
...
+    SET(CMAKE_C_COMPILER "/usr/bin/clang" CACHE PATH "" FORCE)
+    SET(CMAKE_CXX_COMPILER "/usr/bin/clang++" CACHE PATH "" FORCE)
+    add_compile_options(-fsanitize=address,fuzzer-no-link)
+    add_link_options(-fsanitize=address,fuzzer)
+elseif (${AVM_FUZZER} MATCHES "honggfuzz")
+    SET(CMAKE_C_COMPILER "/usr/local/bin/hfuzz-clang" CACHE PATH "" FORCE)
+    SET(CMAKE_CXX_COMPILER "/usr/local/bin/hfuzz-clang++" CACHE PATH "" FORCE)
+    add_compile_options(-fsanitize=address)
+    add_link_options(-fsanitize=address)
+endif ()
```
Comment (hkalodner) [inline packages/arb-avm-cpp/CMakeLists.txt:78]: I don't love the hardcoded paths. Maybe better to move to setting via the command line instead via `cmake -DCMAKE_CXX_COMPILER=afl-clang-lto++ ..` instead? I was looking at https://github.com/jefftrull/json_spirit/blob/develop/fuzzing/CMakeLists.txt for example

File: packages/arb-avm-cpp/avm_values/src/value.cpp [hunk 2]
```diff
@@ -268,7 +269,19 @@ struct ValuePrinter {
...
+        if (b.data_length() <= 64) {
+            os << "0x";
+            std::ios prev_flags(nullptr);
+            prev_flags.copyfmt(os);
+            for (auto b : b.toFlatVector()) {
+                os << std::hex << std::setw(2) << std::setfill('0') << (int)b;
+            }
+            os.copyfmt(prev_flags);
+        } else {
+            os << b.hash();
```
Comment (hkalodner) [inline packages/arb-avm-cpp/avm_values/src/value.cpp:282]: I think either we should just print the data in all cases or we should change the hash output to include the word hash or something to distinguish between a 32 byte buffer and a buffer longer than 64 bytes

File: packages/arb-avm-cpp/fuzz_target/utils.hpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#pragma once
+
+#include <avm/machine.hpp>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+template <class... Args>
+void fuzz_require(bool check, Args... message) {
```
Comment (hkalodner) [inline packages/arb-avm-cpp/fuzz_target/utils.hpp:26]: I doubt performance is an issue here, but you likely want `Args&&...` for perfect forwarding. Also, nice variadic expansion below 😁

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-avm-cpp/avm_values/src/value.cpp (hunk 2 above)
Comment (PlasmaPower) [reply packages/arb-avm-cpp/avm_values/src/value.cpp:282]: Hash isn't 0x prefixed but yeah probably

File: packages/arb-avm-cpp/fuzz_target/utils.hpp (hunk 3 above)
Comment (PlasmaPower) [reply packages/arb-avm-cpp/fuzz_target/utils.hpp:26]: Haha thanks I stole it from stack overflow. I'll fix it.

File: packages/arb-avm-cpp/CMakeLists.txt (hunk 1 above)
Comment (PlasmaPower) [reply packages/arb-avm-cpp/CMakeLists.txt:78]: My goal here was for this to be as easy to setup as possible. Ideally I'd specify a non-absolute path in the CMakeLists.txt but but that doesn't seem possible.

**Review by hkalodner — DISMISSED** [review]
LGTM

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**